src/mapgen/mapgen_valleys.cpp

```diff
@@ -393,18 +393,26 @@ float MapgenValleys::terrainLevelFromNoise(TerrainNoise *tn)
 float MapgenValleys::adjustedTerrainLevelFromNoise(TerrainNoise *tn)
 {
 	float mount = terrainLevelFromNoise(tn);
+	float result = mount;
 	s16 y_start = myround(mount);
 
-	for (s16 y = y_start; y <= y_start + 1000; y++) {
-		float fill =
+	float fill =
+		NoisePerlin3D(&noise_inter_valley_fill->np, tn->x, y_start, tn->z, seed);
+	bool is_ground = fill * *tn->slope >= y_start - mount;
+	s16 search_direction = is_ground ? 1 : -1;
+
+	for (s16 i = 1; i <= 1000; i++) {
+		s16 y = y_start + i * search_direction;
+		fill =
 			NoisePerlin3D(&noise_inter_valley_fill->np, tn->x, y, tn->z, seed);
-		if (fill * *tn->slope < y - mount) {
-			mount = std::fmax((float)(y - 1), mount);
-			break;
-		}
+
+		bool was_ground = is_ground;
+		is_ground = fill * *tn->slope >= y - mount;
+		if (is_ground) result = y;
+		if (is_ground != was_ground) break;
```

Comment:
For both these conditionals the action should be indented on the following line.

Reply:
I'm not sure this will work for when the initial ground level is air, as `result` won't be reduced as this searches downwards?

Reply:
If the initial ground level is air, the result will be reduced upon finding the first ground block.

To avoid the compiler being unable to catch refactoring errors from moving code around or commenting lines out, I normally avoid splitting actions from their conditionals in ways that leave both "half-lines" independently executable. Extra braces (one true brace style) achieve this when a split is necessary. However, consistency is more important, and I now notice the code has been very consistently formatted this way, and that it's even in bold red in the style guide :) 

I'll clean these things up.

Having a separate search loop for each direction and an if statement to select which one to use might have been a more easily readable way of writing it.

Reply:
Yes just seen this will work.